lib/fs.js

```diff
@@ -1639,39 +1650,46 @@ fs.realpathSync = function realpathSync(p, options) {
     pos = nextPartRe.lastIndex;
 
     // continue if not a symlink
-    if (knownHard[base]) {
+    if (knownHard[base] || (cache && cache.get(base) === base)) {
       continue;
     }
 
     var resolvedLink;
-    var stat = fs.lstatSync(base);
-    if (!stat.isSymbolicLink()) {
-      knownHard[base] = true;
-      continue;
-    }
+    const maybeCachedResolved = cache && cache.get(base);
+    if (maybeCachedResolved) {
+      resolvedLink = maybeCachedResolved;
+    } else {
+      var stat = fs.lstatSync(base);
+      if (!stat.isSymbolicLink()) {
+        knownHard[base] = true;
+        continue;
+      }
 
-    // read the link if it wasn't read before
-    // dev/ino always return 0 on windows, so skip the check.
-    var linkTarget = null;
-    if (!isWindows) {
-      var id = stat.dev.toString(32) + ':' + stat.ino.toString(32);
-      if (seenLinks.hasOwnProperty(id)) {
-        linkTarget = seenLinks[id];
+      // read the link if it wasn't read before
+      // dev/ino always return 0 on windows, so skip the check.
+      var linkTarget = null;
+      if (!isWindows) {
+        var id = stat.dev.toString(32) + ':' + stat.ino.toString(32);
```

Comment:
Good idea, done. I’ve also applied this change to the async version to keep them in sync. ;)